### PR TITLE
Add Terratest Example for Compute Module

### PR DESCRIPTION
**Pull request by foreverXZC**
This example uses Terratest to test compute module. All introduction and usages are included in Terratest/README.md.

### CODE REVIEW

**Comment by foreverXZC** [comment]
Removed kitchen terraform and added terratest as travis CI testing.
Trivial things still need to be modified, including readme as well as some pieces of code and configuration.

**Comment by foreverXZC** [comment]
After hiding environment variables, the build fails because the log demonstrates that encrypted environment variables have been removed for security reasons. Although it passes in my own repo, it seems impossible to pass here because pull request from a fork is considered not trustful and thus cannot use encrypted environment variables such as subscription ID, etc.

File: .travis.yml [hunk 1]
```diff
@@ -27,5 +27,5 @@ jobs:
       if: type = push AND branch = master
       install: true
       script:
-      - docker build --build-arg BUILD_TERRAFORM_VERSION=${TERRAFORM_VERSION} --build-arg BUILD_ARM_SUBSCRIPTION_ID=$ARM_SUBSCRIPTION_ID --build-arg BUILD_ARM_CLIENT_ID=$ARM_CLIENT_ID --build-arg BUILD_ARM_CLIENT_SECRET=$ARM_CLIENT_SECRET --build-arg BUILD_ARM_TENANT_ID=$ARM_TENANT_ID -t ${IMAGE_NAME} .
+      - docker build --build-arg BUILD_TERRAFORM_VERSION=${TERRAFORM_VERSION} --build-arg BUILD_ARM_SUBSCRIPTION_ID=$ARM_SUBSCRIPTION_ID --build-arg BUILD_ARM_CLIENT_ID=$ARM_CLIENT_ID --build-arg BUILD_ARM_CLIENT_SECRET=$ARM_CLIENT_SECRET --build-arg BUILD_ARM_TENANT_ID=$ARM_TENANT_ID -t ${IMAGE_NAME} .
```
Comment (metacpp) [inline .travis.yml]: extra white space at the end of line.

File: Dockerfile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Pull the base image with given version.
-ARG BUILD_TERRAFORM_VERSION="0.11.1"
+ARG BUILD_TERRAFORM_VERSION="0.11.3"
```
Comment (metacpp) [inline Dockerfile]: Should be 0.11.7 now.

File: Dockerfile [hunk 3]
```diff
@@ -20,10 +20,25 @@ ENV ARM_TENANT_ID=${BUILD_ARM_TENANT_ID}
...
 
+# Set work directory and generate ssh key
 RUN mkdir /usr/src/${MODULE_NAME}
 COPY . /usr/src/${MODULE_NAME}
-
+WORKDIR /usr/src/${MODULE_NAME}
 RUN ssh-keygen -q -t rsa -b 4096 -f $HOME/.ssh/id_rsa
 
-WORKDIR /usr/src/${MODULE_NAME}
+# Install new version of terraform and golang
```
Comment (metacpp) [inline Dockerfile]: It's already contained in base image, we don't need it any more.

File: Dockerfile [hunk 4]
```diff
@@ -20,10 +20,25 @@ ENV ARM_TENANT_ID=${BUILD_ARM_TENANT_ID}
...
+RUN wget https://releases.hashicorp.com/terraform/0.11.7/terraform_0.11.7_linux_amd64.zip >/dev/null 2>&1
+RUN unzip terraform_0.11.7_linux_amd64.zip >/dev/null
+RUN wget https://storage.googleapis.com/golang/go1.10.3.linux-amd64.tar.gz >/dev/null 2>&1
+RUN tar -zxvf go1.10.3.linux-amd64.tar.gz -C /usr/local/ >/dev/null
+RUN mv terraform /usr/local/bin
+
+# Install required go packages
+ENV GOPATH $HOME/terratest/ssh
```
Comment (metacpp) [inline Dockerfile]: The environment variables related to Go should be set while baking base image.

File: README.md [hunk 5]
```diff
@@ -1,8 +1,8 @@
-# terraform-azurerm-compute #
-[![Build Status](https://travis-ci.org/Azure/terraform-azurerm-compute.svg?branch=master)](https://travis-ci.org/Azure/terraform-azurerm-compute)
+# terraform-azurerm-compute
 
-Deploys 1+ Virtual Machines to your provided VNet
-=================================================
+[![Build Status](https://travis-ci.org/foreverXZC/terraform-azurerm-compute.svg?branch=master)](https://travis-ci.org/foreverXZC/terraform-azurerm-compute)
```
Comment (metacpp) [inline README.md]: The badge should not be yours. Please keep the original one.

File: README.md [hunk 6]
```diff
@@ -170,21 +169,22 @@ We provide 2 ways to build, run, and test the module on a local development mach
...
+curl -sSL https://raw.githubusercontent.com/Azure/terramodtest/master/tool/env_setup.sh | sudo bash
 ```
 
 Then simply run it in local shell:
+
 ```sh
-$ bundle install
-$ rake build
-$ rake e2e
+bundle install
```
Comment (metacpp) [inline README.md]: Why did you change the format here? Can we keep the indention?

File: README.md [hunk 7]
```diff
@@ -170,21 +169,22 @@ We provide 2 ways to build, run, and test the module on a local development mach
...
+rake build
+rake e2e
 ```
 
 ### Docker
 
 We provide a Dockerfile to build a new image based `FROM` the `microsoft/terraform-test` Docker hub image which adds additional tools / packages specific for this module (see Custom Image section).  Alternatively use only the `microsoft/terraform-test` Docker hub image [by using these instructions](https://github.com/Azure/terraform-test).
 
-#### Prerequisites
+#### Prerequisites (Docker)
```
Comment (metacpp) [inline README.md]: It's already under Docker section, don't need to have Docker any more.

File: README.md [hunk 8]
```diff
@@ -161,6 +159,7 @@ We provide 2 ways to build, run, and test the module on a local development mach
 ### Native (Mac/Linux)
 
 #### Prerequisites
+
 - [Ruby **(~> 2.3)**](https://www.ruby-lang.org/en/downloads/)
 - [Bundler **(~> 1.15)**](https://bundler.io/)
 - [Terraform **(~> 0.11.0)**](https://www.terraform.io/downloads.html)
```
Comment (metacpp) [inline README.md]: 0.11.7

Comment (metacpp) [inline README.md]: Go SDK.

File: README.md [hunk 9]
```diff
@@ -170,21 +169,22 @@ We provide 2 ways to build, run, and test the module on a local development mach
 We provide simple script to quickly set up module development environment:
 
 ```sh
-$ curl -sSL https://raw.githubusercontent.com/Azure/terramodtest/master/tool/env_setup.sh | sudo bash
+curl -sSL https://raw.githubusercontent.com/Azure/terramodtest/master/tool/env_setup.sh | sudo bash
```
Comment (metacpp) [inline README.md]: Did you verify if this script needs to be updated for new testing environment?

File: Rakefile [hunk 10]
```diff
@@ -26,21 +19,15 @@ namespace :static do
...
-  end
-  task :verify do
-    kitchen_verify
-  end
   task :test do
-    kitchen_test
-  end
-  task :destroy do
-    kitchen_destroy
+    success = system ("go test -v terratest/ssh/terraform_ssh_example_test.go -timeout 20m -args azureuser ~/.ssh/id_rsa")
```
Comment (metacpp) [inline Rakefile]: test_folder_name/**/*_test.go?

File: .travis.yml (hunk 1 above)
Comment (foreverXZC) [reply .travis.yml]: Will delete.

File: Dockerfile (hunk 2 above)
Comment (foreverXZC) [reply Dockerfile]: However, 0.11.7 is not released in docker hub, which means if we set terraform version as 0.11.7 here, an error occurs showing that no such kind of docker image.

File: README.md (hunk 5 above)
Comment (foreverXZC) [reply README.md]: Will change.

File: README.md (hunk 6 above)
Comment (foreverXZC) [reply README.md]: VS Code thinks it is not a good idea to have such format in markdown. It illustrates "dollar signs used before commands without showing output". See: https://github.com/updownpress/markdown-lint/blob/master/rules/014-commands-show-output.md

File: README.md (hunk 7 above)
Comment (foreverXZC) [reply README.md]: VS Code considers it as the same heading as the previous one. Thus it thinks the heading is duplicated.

File: README.md (hunk 9 above)
Comment (foreverXZC) [reply README.md]: Yes the script should be modified if we introduce Go SDK.

File: Rakefile (hunk 10 above)
Comment (foreverXZC) [reply Rakefile]: It seems that this format using * does not work...

File: terratest/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
+# Compute Example
```
Comment (metacpp) [inline terratest/README.md]: Please merge it with README in root folder, remove unnecessary content like how to install go and terraform.

File: terratest/compute/main.tf [hunk 12]
```diff
@@ -0,0 +1,41 @@
+provider "random" {
```
Comment (metacpp) [inline terratest/compute/main.tf]: Put all the tf code for testing into ./test/fixture

File: terratest/ssh/public_key_file_deprecated.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package test
+
+import (
+	"io/ioutil"
```
Comment (metacpp) [inline terratest/ssh/public_key_file_deprecated.go]: Use  2 whitespaces as indention.

**Comment by metacpp** [comment]
@foreverXZC please remove the .testdata folder and add it into .gitignore.

File: Dockerfile (hunk 3 above)
Comment (foreverXZC) [reply Dockerfile]: Will remove.

File: Dockerfile (hunk 4 above)
Comment (foreverXZC) [reply Dockerfile]: Will put in base image.

File: terratest/README.md (hunk 11 above)
Comment (foreverXZC) [reply terratest/README.md]: Yes after I ensure that the script runs correctly,

File: terratest/compute/main.tf (hunk 12 above)
Comment (foreverXZC) [reply terratest/compute/main.tf]: Will change.

File: terratest/ssh/public_key_file_deprecated.go (hunk 13 above)
Comment (foreverXZC) [reply terratest/ssh/public_key_file_deprecated.go]: OK will modify. Nevertheless, although most programming languages use a tab or four white spaces as indentation, terraform format uses two white spaces, which seems quite interesting but a bit weird.

File: README.md (hunk 9 above)
Comment (foreverXZC) [reply README.md]: Unfortunately, "export" commands in script only affects environment variables inside the script, which means users have to set global environment variables by themselves. Will add export commands here.

File: README.md (hunk 6 above)
Comment (foreverXZC) [reply README.md]: In addition, without docker, users have to install go packages manually, which cannot be put in env.sh because they are module related.

File: Dockerfile (hunk 2 above)
Comment (metacpp) [reply Dockerfile]: 0.11.7 was pushed, go ahead to use it.

File: Dockerfile (hunk 3 above)
Comment (metacpp) [reply Dockerfile]: Can you remove them?

File: Dockerfile [hunk 14]
```diff
@@ -20,10 +20,25 @@ ENV ARM_TENANT_ID=${BUILD_ARM_TENANT_ID}
...
-WORKDIR /usr/src/${MODULE_NAME}
+# Install new version of terraform and golang
+RUN wget https://releases.hashicorp.com/terraform/0.11.7/terraform_0.11.7_linux_amd64.zip >/dev/null 2>&1
+RUN unzip terraform_0.11.7_linux_amd64.zip >/dev/null
+RUN wget https://storage.googleapis.com/golang/go1.10.3.linux-amd64.tar.gz >/dev/null 2>&1
+RUN tar -zxvf go1.10.3.linux-amd64.tar.gz -C /usr/local/ >/dev/null
+RUN mv terraform /usr/local/bin
+
+# Install required go packages
+ENV GOPATH $HOME/go
```
Comment (metacpp) [inline Dockerfile]: You should remove the steps contained in base image.

File: Dockerfile [hunk 15]
```diff
@@ -20,10 +20,25 @@ ENV ARM_TENANT_ID=${BUILD_ARM_TENANT_ID}
...
+ENV PATH /usr/local/go/bin:$PATH
+RUN /bin/bash -c "go get github.com/gruntwork-io/terratest/modules/ssh"
```
Comment (metacpp) [inline Dockerfile]: these pkgs should be restored through go dep.

File: README.md (hunk 5 above)
Comment (metacpp) [reply README.md]: Can you please change it?

File: README.md [hunk 16]
```diff
@@ -173,7 +173,16 @@ We provide simple script to quickly set up module development environment:
 $ curl -sSL https://raw.githubusercontent.com/Azure/terramodtest/master/tool/env_setup.sh | sudo bash
 ```
 
-Then simply run it in local shell:
+Then use dep ensure to install go packages:
+
+```sh
+$ cd $GOPATH/src/{directory_name}/test/compute
```
Comment (metacpp) [inline README.md]: I don't think you need to specifically do this. All the steps should be supposed to run at the root folder of each module project.

File: README.md [hunk 17]
```diff
@@ -173,7 +173,16 @@ We provide simple script to quickly set up module development environment:
...
+$ dep ensure
+$ cd ../..
+```
+
+Finally simply run it in local shell:
+
 ```sh
 $ bundle install
 $ rake build
```
Comment (metacpp) [inline README.md]: Why don't you make `make dep` into task of rake build system?

File: Rakefile [hunk 18]
```diff
@@ -6,13 +6,6 @@ require 'rspec/core/rake_task'
 require 'bundler/setup'
 require 'terramodtest'
 
-namespace :presteps do
```
Comment (metacpp) [inline Rakefile]: You may still keep presteps task to run `dep restore`.

File: Rakefile [hunk 19]
```diff
@@ -26,21 +19,15 @@ namespace :static do
...
-  end
-  task :verify do
-    kitchen_verify
-  end
   task :test do
-    kitchen_test
-  end
-  task :destroy do
-    kitchen_destroy
+    success = system ("go test -v ./test/compute/ -timeout 20m -args azureuser ~/.ssh/id_rsa")
```
Comment (metacpp) [inline Rakefile]: ./test should be enough here, compute subfolder is meaningless in compute project.

File: test/README.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
+# Compute Example
```
Comment (metacpp) [inline test/README.md]: Why did you still keep this README? You should remove merge the testing related part into README.md under root folder.

File: test/compute/Gopkg.lock [hunk 21]
```diff
@@ -0,0 +1,158 @@
+# This file is autogenerated, do not edit; changes may be undone by the next 'dep ensure'.
```
Comment (metacpp) [inline test/compute/Gopkg.lock]: Please don't use `compute` folder under test folder, this project is for compute only.

File: test/compute/Gopkg.lock [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+
+
```
Comment (metacpp) [inline test/compute/Gopkg.lock]: Gopkg.lock and Gopkg.toml can be directly put under root folder.

File: .gitignore [hunk 23]
```diff
@@ -9,7 +9,7 @@ terraform.tfvars
 
 # Module directory
 .terraform/
-
+.test-data/
```
Comment (metacpp) [inline .gitignore]: Missed vendor folder.

File: Rakefile (hunk 19 above)
Comment (foreverXZC) [reply Rakefile]: OK I will move these go files outside.

File: test/README.md (hunk 20 above)
Comment (foreverXZC) [reply test/README.md]: The main README seems enough, so now I think this one can be removed.

File: test/compute/Gopkg.lock (hunk 21 above)
Comment (foreverXZC) [reply test/compute/Gopkg.lock]: Yes will move go files outside.

File: test/compute/Gopkg.lock (hunk 22 above)
Comment (foreverXZC) [reply test/compute/Gopkg.lock]: Makes sense.

File: .gitignore (hunk 23 above)
Comment (foreverXZC) [reply .gitignore]: Will add.

File: README.md (hunk 16 above)
Comment (foreverXZC) [reply README.md]: Yes I will put Gopkg.* in root.

File: README.md (hunk 17 above)
Comment (foreverXZC) [reply README.md]: If it works, will merge dep into Rakefile.

File: Rakefile (hunk 18 above)
Comment (foreverXZC) [reply Rakefile]: Will try.

File: Dockerfile (hunk 3 above)
Comment (metacpp) [reply Dockerfile]: It does not matter coz I will not merge it until you verified everything from E2E.

File: README.md (hunk 5 above)
Comment (metacpp) [reply README.md]: Validation should always happen after you make any change. And merge will never happen if you don't do any validation.

File: README.md (hunk 9 above)
Comment (metacpp) [reply README.md]: If that's the case, you can try `source` command in linux.

File: terratest/ssh/public_key_file_deprecated.go (hunk 13 above)
Comment (metacpp) [reply terratest/ssh/public_key_file_deprecated.go]: That's not true. Many languages(C++, JS) uses 2 whitespaces. Anyway, let's be consistent here.

File: Dockerfile (hunk 15 above)
Comment (metacpp) [reply Dockerfile]: You can install `dep` in this Dockerfile, that's OK. But you should remove the Go packages in Dockerfile, they should be restored by 'dep ensure'.

**Comment by metacpp** [comment]
@foreverXZC let me know if e2e validation is successful.
